src/bls12_381/ec.rs

```diff
@@ -617,10 +695,40 @@ macro_rules! curve_impl {
                 }
             }
         }
+
+        #[cfg(test)]
+        use rand::{SeedableRng, XorShiftRng};
+
+        #[test]
+        fn test_hash() {
+            let mut rng = XorShiftRng::from_seed([0x5dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+
+            let seed : [u8; 32] = rng.gen();
+            let p = $affine::hash(&seed);
+            assert!(!p.is_zero());
+            assert!(p.is_on_curve());
+            assert!(p.is_in_correct_subgroup_assuming_on_curve());
+        }
+
+        #[test]
+        fn test_sw_encode() {
+            let mut rng = XorShiftRng::from_seed([0x5dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+
+            for _ in 0 .. 20 {
+                let t = $basefield::rand(&mut rng);
+                let p = $affine::sw_encode(t);
+                assert!(p.is_on_curve());
+                assert!(!p.is_zero());
+            }
+
+            let p = $affine::sw_encode($basefield::zero());
+            assert!(p.is_on_curve());
```

Comment:
We should also test that the two "exceptional points" `t^2 + 1 + b = 0` return the correct points here. (You can obtain these points generically by taking sqrt of `- 1 - coeff_b()`. These points only exist for G1, but don't exist for G2 -- let's test that too.